### DISCUSSION TITLE
custom image "loader"

### DISCUSSION THREAD

**Discussion by farfromrefug** [SDWebImage/SDWebImage#3754] [Q&A] [answered]
I have a project where some image are 4bit bmp images.
Right now it will fail because UIImage does know how to load 4bit bitmap.
I do have some code to create UIImage from 4bit bitmap.
Is there any way in sdwebimage to register a custom image "loader" so that i can add my custom code?

**Comment by dreampiggy** [comment, marked as answer]
See: https://github.com/SDWebImage/SDWebImage/wiki/Coder-Plugin-List
API is simple, just implements the `SDImageCoder` protocol, which 4 is required (2 for decoding and 2 for encoding, if you don't care about encoding, just return nil)

Demo: https://github.com/SDWebImage/SDWebImageProgressiveJPEGDemo

**Reply by farfromrefug** [reply]
@dreampiggy awesome! and the "result" is cached right? memory / disk ?
thanks

**Reply by dreampiggy** [reply]
Image Coder is just decoding part. Other part does not get effected and you (Image Decoder author) don't need to care about

**Reply by farfromrefug** [reply]
@dreampiggy thanks i just wanted to make sure that once decoded/chached the image picked up from cache the next time does not need that decoding anymore ? 
I mean is the cached image the decoded one or the "source" one.
Thanks again for everything

**Reply by dreampiggy** [reply]
> once decoded/chached the image picked up from cache the next time does not need that decoding anymore

Yes, for simple cases. Let me explain.

Assume your coder introduce a new image format we never see (like `MYFORMAT`)

SD's cache use memory and disk cache. For in-memory , we cache the `UIImage` (which means, your Image Coder returned decoded image object). So it's OK and we move this image object around.

For disk cache, since we need a way to `serialize` the `UIImage` into `Data` and write to disk, if you don't implements `encodedData` related encoding protocol API, normally we will cache the original downloaded data buffer into disk.

But some advanced feature like [Transformer](https://github.com/SDWebImage/SDWebImage/wiki/Advanced-Usage#image-transformer-50) (like `SDImageTintTransformer`, which tint the original decoded image into **another different one** with color), this will fallback to use JPEG/PNG to serialize to the `Data`. 

Because your coder don't support encoding, and we don't have the same data matching the new transformed `UIImage`, so this will use JPEG/PNG as backup solution. So in conclusion:

1. Download `1.MYFORMAT` data, it stored into the disk cache for key `https://1.MYFORMAT`
2. The in-memory UIImage stored into memory cache for `https://1.MYFORMAT`
3. When use transformer, it generate a new cache key for `https://1.MYFORMAT-transformer(Foobar)`, and store the transformed image to memory cache
4. It encode the transformed image into JPEG (instead of `MYFORMAT`) and stored to disk cache for key `https://1.MYFORMAT-transformer(Foobar)`

Can my explanation solve your question ?

**Reply by farfromrefug** [reply]
@dreampiggy yes I think it does explains it perfectly! So if I implement the encodedData(which I can as the decoded image is a simple bitmap) it - should work best and only decode "once" then always use the cache "decoded" image.
Thanks a lot really love what you Ve created!

**Reply by dreampiggy** [reply]
It's recommended to implements the encoding part. Most of image codec supports encoding. Which can make sure we always use your format instead of transcoding to another format like JPEG/PNG which may loss some information inside the image format, like EXIF
